packages/concerto-core/lib/introspect/mapvaluetype.js

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+const ModelUtil = require('../modelutil');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapValueType defines a Value type of MapDeclaration.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapValueType extends Decorated {
+    /**
+     * Create an MapValueType.
+     * @param {MapDeclaration} parent - The owner of this property
+     * @param {Object} ast - The AST created by the parser
+     * @throws {IllegalModelException}
+     */
+    constructor(parent, ast) {
+        super(ast);
+        this.parent = parent;
+        this.name = null;
+        this.type = null;
+        this.fqn = null;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+        this.name = this.ast.name;
+        this.type = this.ast.name;
+        this.fqn = ModelUtil.getFullyQualifiedName(this.parent.getModelFile().getNamespace(), this.ast.name);
+    }
+
+    /**
+     * Semantic validation of the structure of this class.
+     *
+     * @throws {IllegalModelException}
+     * @protected
+     */
+    validate() {
+        const declarations = this.getModelFile().getAllDeclarations();
+
+        const value = declarations.find(decl => decl.name === this.name);
+
+        if (!value?.isConcept?.()           &&
+            !value?.isEnum?.()              &&
+            !value?.isEvent?.()             &&
```

Comment:
Why do we need to call out Event specifically?

If we do, we should check for Transactions, Assets and Participants too

Reply:
`Event` is a legal map value type , `Transactions`, `Assets` and `Participants` are not specified as value types.

Reply:
Ok, I think that they should. However, Event, Transactions, Assets and Participants are all Concepts IIRC. so this line is redundant?

Reply:
The evaluation expression is based on the `$class` value, which is different for both ConceptDeclaration and EventDeclaration:

`return this.type === ${MetaModelNamespace}.ConceptDeclaration
`

`return this.type === ${MetaModelNamespace}.EventDeclaration
`